cloudmarker/events/vmdatadiskencryptionevent.py

```diff
@@ -0,0 +1,122 @@
+identifies an unencrypted VM data disk. This plugin works on the virtual
+machine properties found in the ``com`` bucket of ``virtual_machine``
+records.
+"""
+
+
+import logging
+
+from cloudmarker import util
+
+_log = logging.getLogger(__name__)
+
+
+class VMDataDiskEncryptionEvent:
+    """VM Data disk encryption event plugin."""
+
+    def __init__(self):
+        """Create an instance of :class:`VMDataDiskEncryptionEvent`."""
+
+    # pylint: disable=R0201
+    def eval(self, record):
+        """Evaluate virtual machine to check for unencrypted data disks.
+
+        Arguments:
+            record (dict): A virtual machine record.
+
+        Yields:
+            dict: An event record representing an unencrypted data disk
+            of a virtual machine
+
+        """
+        # If 'com' bucket is missing, we have a malformed record. Log a
+        # warning and ignore it.
+        com = record.get('com')
+        if com is None:
+            _log.warning('Virtual machine record is missing \'com\' key: %r',
+                         record)
+            return
+
+        # This plugin understands compute rule records only,
+        # so ignore  any other record types.
+        common_record_type = com.get('record_type')
+
+        if common_record_type != 'compute':
+            return
+
+        # If 'ext' bucket is missing, we have a malformed record. Log a
+        # warning and ignore it.
+        ext = record.get('ext')
+        if ext is None:
+            _log.warning('Virtual machine record is missing \'ext\' key: %r',
+                         record)
+            return
+        all_data_disks_encrypted = ext.get('all_data_disks_encrypted')
+
+        if all_data_disks_encrypted:
+            return
+        if com.get('cloud_type') == 'azure':
+            yield from _get_azure_vm_data_disk_encryption_event(
+                com, ext, record.get('raw'))
+
+    def done(self):
+        """Perform cleanup work.
+
+        Currently, this method does nothing. This may change in future.
+        """
+
+
+def _get_azure_vm_data_disk_encryption_event(com, ext, raw):
+    """Evaluate Azure VM for unencrypted data disks.
+
+    Arguments:
+        com (dict): Virtual Machine record `com` bucket
+        ext (dict): Virtual Machine record `ext` bucket
+        raw (dict): Virtual Machine record `raw` bucket
+    Returns:
+        dict: An event record representing unencrypted data disk
+
+    """
+    friendly_cloud_type = util.friendly_string(com.get('cloud_type'))
+    os_disk_name = raw.get('storage_profile').get('os_disk').get('name')
+    instance_view = raw.get('instance_view')
+    if instance_view is None:
+        return
+    for disk in instance_view.get('disks'):
+        # If it is an OS disk skip and continue
+        if disk.get('name') == os_disk_name:
+            continue
+        if disk.get('encryption_settings') is not None and \
+                disk['encryption_settings'][0]['enabled']:
```

Comment:
Instead of checking only `disk['encryption_settings'][0]` here, we might need to loop over all items in `disk['encryption_settings']` and then check each item's `'enabled'` key. I am only guessing here. You are the expert in this area, so you would be able to take the best decision here.

If you also think we need to loop here, it might be useful to create a separate function to do the looping work to prevent this part of the code from becoming more deeply nested.

Reply:
Having tried multiple scenarios I haven't seen a case where the size of `disk['encryption_settings']` is more than 1. I plan to keep it as it, till the time it breaks.